src/internal/Subscription.ts

```diff
@@ -53,180 +61,188 @@ export class Subscription implements SubscriptionLike {
-   * added to the inner subscriptions list. This Subscription can be used with
-   * `remove()` to remove the passed teardown logic from the inner subscriptions
-   * list.
+   * @param teardown The teardown logic to add to this subscription.
    */
   add(teardown: TeardownLogic): void {
-    let subscription = (<Subscription>teardown);
-
-    if (!teardown) {
-      return;
-    }
-
-    switch (typeof teardown) {
-      case 'function':
-        subscription = new Subscription(<(() => void)>teardown);
-      case 'object':
-        if (subscription === this || subscription.closed || typeof subscription.unsubscribe !== 'function') {
-          // This also covers the case where `subscription` is `Subscription.EMPTY`, which is always in `closed` state.
-          return;
-        } else if (this.closed) {
-          subscription.unsubscribe();
-        } else if (!(subscription instanceof Subscription)) {
-          const tmp = subscription;
-          subscription = new Subscription();
-          subscription._subscriptions = [tmp];
+    // Only add the teardown if it's not undefined
+    // and don't add a subscription to itself.
+    if (teardown && teardown !== this) {
+      if (this.closed) {
+        // If this subscription is already closed,
+        // execute whatever teardown is handed to it automatically.
+        if (typeof teardown === 'function') {
+          teardown();
+        } else {
+          teardown.unsubscribe();
         }
-        break;
-      default: {
-        throw new Error('unrecognized teardown ' + teardown + ' added to Subscription.');
+      } else {
+        this._teardowns = this._teardowns ?? [];
+        if (teardown instanceof Subscription) {
+          // We don't add closed subscriptions, and we don't add the same subscription
+          // twice. Subscription unsubscribe is idempotent.
+          if (teardown.closed || teardown._hasParent(this)) {
+            return;
+          }
+          teardown._addParent(this);
+        }
+        this._teardowns.push(teardown);
       }
     }
+  }
+
+  private _hasParent(parent: Subscription) {
+    return this._singleParent === parent || this._parents?.includes(parent) || false;
+  }
 
-    // Add `this` as parent of `subscription` if that's not already the case.
-    let { _parentOrParents } = subscription;
-    if (_parentOrParents == null) {
-      // If we don't have a parent, then set `subscription._parents` to
-      // the `this`, which is the common case that we optimize for.
-      subscription._parentOrParents = this;
-    } else if (_parentOrParents instanceof Subscription) {
-      if (_parentOrParents === this) {
-        // The `subscription` already has `this` as a parent.
-        return;
+  private _addParent(parent: Subscription) {
+    // If we already have one or more parents, we need
+    // to add to the list of parents.
+    if ((this._parents || this._singleParent) && this._singleParent !== parent) {
```

Comment:
nitpick: For me, the parentheses in this condition make it more difficult to understand. To me, it would be easier to understand as `this._parents || (this._singleParent && this._singleParent !== parent)`. Prettier will likely remove the parentheses, as they are not necessary because of operator precedence, but putting them around `this._parents || this._singleParent` doesn't make this easier to understand, IMO.

Reply:
I must have been tired. I don't like any of of this method now... Refactored completely to be more readable and less nonsense. Those `_parents!` checks were a smell.